api/v1alpha3/machinedeployment_webhook.go

```diff
@@ -78,3 +86,54 @@ func (m *MachineDeployment) validate() error {
 
 	return apierrors.NewInvalid(GroupVersion.WithKind("MachineDeployment").GroupKind(), m.Name, allErrs)
 }
+
+// PopulateDefaultsMachineDeployment fills in default field values
+// This is also called during MachineDeployment sync.
+func PopulateDefaultsMachineDeployment(d *MachineDeployment) {
+	if d.Spec.Replicas == nil {
+		d.Spec.Replicas = new(int32)
+		*d.Spec.Replicas = 1
+	}
+
+	if d.Spec.MinReadySeconds == nil {
+		d.Spec.MinReadySeconds = new(int32)
+		*d.Spec.MinReadySeconds = 0
+	}
+
+	if d.Spec.RevisionHistoryLimit == nil {
+		d.Spec.RevisionHistoryLimit = new(int32)
+		*d.Spec.RevisionHistoryLimit = 1
+	}
+
+	if d.Spec.ProgressDeadlineSeconds == nil {
+		d.Spec.ProgressDeadlineSeconds = new(int32)
+		*d.Spec.ProgressDeadlineSeconds = 600
+	}
+
+	if d.Spec.Strategy == nil {
+		d.Spec.Strategy = &MachineDeploymentStrategy{}
+	}
+
+	if d.Spec.Strategy.Type == "" {
+		d.Spec.Strategy.Type = RollingUpdateMachineDeploymentStrategyType
+	}
+
+	// Default RollingUpdate strategy only if strategy type is RollingUpdate.
+	if d.Spec.Strategy.Type == RollingUpdateMachineDeploymentStrategyType {
+		if d.Spec.Strategy.RollingUpdate == nil {
+			d.Spec.Strategy.RollingUpdate = &MachineRollingUpdateDeployment{}
+		}
+		if d.Spec.Strategy.RollingUpdate.MaxSurge == nil {
+			ios1 := intstr.FromInt(1)
+			d.Spec.Strategy.RollingUpdate.MaxSurge = &ios1
+		}
+		if d.Spec.Strategy.RollingUpdate.MaxUnavailable == nil {
+			ios0 := intstr.FromInt(0)
+			d.Spec.Strategy.RollingUpdate.MaxUnavailable = &ios0
+		}
+	}
+
+	if len(d.Namespace) == 0 {
```

Comment:
I noticed this specific behavior here. Is this required as part of the machinedeployment sync?

I did not add this behavior to the other objects because as part of my manual testing using `kubectl apply`, the namespace on the Cluster and Machine were being added before reaching the mutating webhook logic even if it wasn't specified in yaml.

Reply:
The namespace should always be set when the resource is being provided by the api server (or the controller-runtime cache). If we are building up a new MachineDeployment from code during the sync process, then I suspect it may be needed.

Reply:
We aren't creating new MDs and then calling this. I would remove this.